### PR TITLE
Use reflection or a compile-time only shim assembly to reference unexposed corelib types.

### PR DESCRIPTION
**Pull request by jkoritzinsky**
Use reflection for the hosting tests as they'll be around indefinitely.

Use the ref-assembly trick for ICastable testing since ICastable testing would require Ref-emit which makes the test significantly less readable and since ICastable will be going away within the .NET 7 timeframe. Supercedes #61754

### CODE REVIEW

File: src/tests/Interop/COM/Activator/ComActivationContextShim.cs [hunk 1]
```diff
@@ -0,0 +1,46 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Reflection;
+
+namespace Activator
+{
+    internal sealed class ComActivationContextShim
```
Comment (jkotas) [inline src/tests/Interop/COM/Activator/ComActivationContextShim.cs:9]: Nit: You can create a local `ComActivationContext` definition, get the pointer to the method and invoke it via the pointer. The types do not have to match that way, and it better matches how the method is called in the real shim.

File: src/tests/Interop/ICastable/ICastable.CoreLib.csproj [hunk 2]
```diff
@@ -0,0 +1,17 @@
+<Project Sdk="Microsoft.NET.Sdk">
+  <!--
+    ICastable.CoreLib provides a "shim" definition of ICastable in System.Private.CoreLib for the Castable test suite.
```
Comment (jkotas) [inline src/tests/Interop/ICastable/ICastable.CoreLib.csproj:3]: Neat hack :-)

**Review by jkotas — APPROVED** [review]
LGTM. Thank you!

**Review by AaronRobinsonMSFT — APPROVED** [review]
This is works too.

My opinion here is that this sort of code is much harder to work with and is far more annoying to use than simply having a pseudo ref assembly. The shim/proxy type approach works but it does make things much harder to debug in managed code.

Thanks for removing the SPCL ref logic, that is a win either way.

**Comment by jkoritzinsky** [comment]
Looks like things break really badly if CoreLib has any types marked with `[TypeIdentifierAttribute]`. I'll rewrite the Activator test to do something else to still be able to test things.

File: src/tests/Interop/COM/Activator/Activator.csproj [hunk 3]
```diff
@@ -18,7 +19,7 @@
     <ProjectReference Include="Servers\AssemblyContracts.csproj" />
   </ItemGroup>
   <!-- Set up an isolated directory for the server assemblies -->
-  <Target Name="SetUpIsolatedDirectory" AfterTargets="CopyFilesToOutputDirectory">
+  <Target Name="SetUpIslatedDirectory" AfterTargets="CopyFilesToOutputDirectory">
```
Comment (AaronRobinsonMSFT) [inline src/tests/Interop/COM/Activator/Activator.csproj:22]:
```suggestion
  <Target Name="SetUpIsolatedDirectory" AfterTargets="CopyFilesToOutputDirectory">
```

**Review by AaronRobinsonMSFT — APPROVED** [review]

**Pull request merged**